File: frontend/src/types/trip.ts
export interface TripLeg {
  id: string;
  sequence: number;
  origin_airport: string;
  origin_city: string;
  destination_airport: string;
  destination_city: string;
  preferred_date: string;
  flexibility_days: number;
  cabin_class: string;
  passengers: number;
  companion_preferred_date: string | null;
}

export interface Trip {
  id: string;
  title: string | null;
  status: string;
  natural_language_input: string | null;
  parsed_input: Record<string, unknown> | null;
  total_estimated_cost: number | null;
  currency: string;
  companions: number;
  companion_cabin_class: string | null;
  legs: TripLeg[];
  created_at: string;
  updated_at: string;
}

export interface CompanionLegOption {
  leg_id: string;
  route: string;
  date: string;
  cabin_class: string;
  airline_code: string;
  airline_name: string;
  per_person: number;
  total: number;
  stops: number;
  duration_minutes: number;
}

export interface NearbyDateOption {
  leg_id: string;
  route: string;
  date: string;
  cabin_class: string;
  airline_code: string;
  airline_name: string;
  per_person: number;
  total: number;
  stops: number;
  date_diff_days: number;
  savings_vs_selected: number;
}

export interface CompanionPricing {
  employee_total: number;
  companions_count: number;
  companion_cabin_class: string;
  companion_options: CompanionLegOption[];
  nearby_date_options: NearbyDateOption[];
  combined_min: number;
  combined_max: number;
  summary: string;
}

export interface CabinOption {
  cabin_class: string;
  per_person_per_leg: number[];
  total_per_person: number;
  total_all_travelers: number;
  fits_budget: boolean;
  budget_delta: number;
  budget_delta_percent: number;
  airline_codes: string[];
}

export interface CabinBudgetResult {
  anchor_total: number;
  budget_envelope: number;
  budget_tolerance: number;
  total_travelers: number;
  recommended_cabin: string;
  recommendation_reason: string;
  cabin_options: CabinOption[];
  economy_savings: number;
  near_miss_note?: string | null;
  savings_note?: string | null;
  source: "llm" | "fallback";
}

File: backend/app/schemas/trip.py
import uuid
from datetime import date, datetime

from pydantic import BaseModel


class TripLegBase(BaseModel):
    origin_city: str
    destination_city: str
    preferred_date: date
    flexibility_days: int = 3
    cabin_class: str = "economy"
    passengers: int = 1


class TripLegResponse(BaseModel):
    id: uuid.UUID
    sequence: int
    origin_airport: str
    origin_city: str
    destination_airport: str
    destination_city: str
    preferred_date: date
    flexibility_days: int
    cabin_class: str
    passengers: int

    model_config = {"from_attributes": True}


class CreateTripNL(BaseModel):
    natural_language_input: str


class CreateTripStructured(BaseModel):
    legs: list[TripLegBase]


class TripResponse(BaseModel):
    id: uuid.UUID
    title: str | None
    status: str
    natural_language_input: str | None
    parsed_input: dict | None
    total_estimated_cost: float | None
    currency: str
    legs: list[TripLegResponse]
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class UpdateLegsRequest(BaseModel):
    legs: list[dict]


class PatchLegRequest(BaseModel):
    cabin_class: str | None = None
    passengers: int | None = None
    flexibility_days: int | None = None
